qcodes/instrument/mockers/simulated_ats_api.py

```diff
@@ -0,0 +1,112 @@
+
+
+class SimulatedAlazarATSAPI(AlazarATSAPI):
+
+    registers = {
+        8: 70254688,
+        58: np.uint32(1 << 26)  # Trigger hold off
+    }
+
+    def __init__(
+            self,
+            dll_path: str = None,  # Need this for meta super class
+            buffer_generator: Optional[Callable[[np.ndarray], None]] = None,
+            dtype: type = np.uint16
+    ):
+        # don't call `super().__init__` here to avoid dependence on the
+        # alazar driver, when loading the dll
+        self._buffer_generator = (
+            buffer_generator or self._default_buffer_generator)
+        self.dtype = dtype
+        self.buffers = {}
+
+    def _sync_dll_call(self, c_name: str, *args):
+        _mark_params_as_updated(*args)
+
+    def get_board_by_system_id(self, system_id: int, board_id: int) -> int:
+        # This is the handle....
+        return 12335
+
+    def busy(self, handle: int) -> int:
+        return 0
+
+    # OTHER API-RELATED METHODS
+
+    def get_board_model(self, handle: int) -> str:
+        return 'ATS9360'
+
+    def get_channel_info_(self, handle: int) -> Tuple[int, int]:
+        return 4294967294, 12
+
+    def get_cpld_version_(self, handle: int) -> str:
+        return "25.16"
+
+    def get_driver_version_(self) -> str:
+        return '6.5.1'
+
+    def get_sdk_version_(self) -> str:
+        return '6.5.1'
+
+    def query_capability_(self, handle: int, capability: int) -> int:
+        capabilities = {
+            Capability.ASOPC_TYPE: 1763017568,
+            Capability.GET_SERIAL_NUMBER: 970396,
+            Capability.MEMORY_SIZE: 4294967294,
+            Capability.GET_PCIE_LINK_WIDTH: 8,
+            Capability.GET_PCIE_LINK_SPEED: 2,
+            Capability.GET_LATEST_CAL_DATE: 250117}
+        return capabilities[Capability(capability)]
+
+    def read_register_(self, handle: int, offset: int) -> int:
+        return self.registers[offset]
+
+    def write_register_(self, handle: int, offset: int, value: int) -> None:
+        self.registers[offset] = value
+
+    def post_async_buffer(
+            self,
+            handle: int,
+            buffer: ctypes.c_void_p,
+            buffer_length: int
+    ) -> ReturnCode:
+        ctypes_array = (ctypes.c_uint16 *
+                        (buffer_length // 2)).from_address(buffer.value)
+        buf = np.frombuffer(ctypes_array, dtype=self.dtype)
+        self.buffers[buffer.value] = buf
+        return self._sync_dll_call(
+            'AlazarPostAsyncBuffer', handle, buffer, buffer_length)
+
+    def wait_async_buffer_complete(
+            self,
+            handle: int,
+            buffer: ctypes.c_void_p,
+            timeout_in_ms: int
+    ) -> ReturnCode:
+        b = self.buffers.get(buffer.value, None)
+        assert b is not None
+        self._buffer_generator(b)
+        return self._sync_dll_call(
+            'AlazarWaitAsyncBufferComplete', handle, buffer, timeout_in_ms)
+
+    @staticmethod
```

Comment:
why is it a static method and not just a function?

Reply:
This is probably a matter of taste. 
My thought was: it only ever gets used from within the class, and to improve readability, I like to define things on the most restrictive scope possible. Following that idea it should be a local function inside the constructor.
What are your thoughts? Why should this live on a module level?

Reply:
> define things on the most restrictive scope possible

that's perfect, i agree with that. i don't like local function because it's impossible to test them by themselves.

my general thoughts were that the buffer generator is an interface kind of thing that the simulator uses (or expects from outside) rather than something that it "is capable of providing". Hence providing a default from outside of the class sounds logical as opposed to have the default defined inside the class.

ok. since it's anyway 'private' to this class/module, let's keep it as it. but big thanks for explaining the thoughts!